keps/sig-node/753-sidecar-containers/README.md

```diff
@@ -843,24 +843,53 @@ examples:
 
 ### Termination of containers
 
-In alpha sidecar containers will be terminated as regular containers. No special
+In Alpha sidecar containers will be terminated as regular containers. No special
 or additional signals will be supported.
 
-We will collect feedback from alpha implementation and plan to improve
-termination in Beta. When Pods with sidecars are terminated:
+In Beta we have thought about introducing additional termination grace period fields
+to manage termination duration
+([draft proposal](https://docs.google.com/document/d/1B01EdgWJAfkT3l6CIwNwoskQ71eyuwet3mjiQrMQbU8))
+and leverage these fields to add reverse order termination of sidecar containers
+after the primary containers terminate.
 
+However, we decided on an alternative that doesn't require additional fields or hooks while keeping
+the desired behaviors when Pods with sidecars are terminated. While original approach works better
+with truly graceful termination where consistency is more important than time taken, proposed approach
+works for that scenario as well as a more and more popular scenario of limited time to terminate when
+graceful termination is set by external requirement and Pods needs to do best to gracefully terminate
+as much as possible (think of a Spot Instances with 30 seconds notification).
+
+Here is the proposed approach:
+1. Sidecar containers that have a `PreStop` hook will be notified when the Pod has begun terminating
+   by executing the `PreStop` hook. This happens at the same time as regular containers, and begins
+   the Pod's termination grace period countdown.
+2. Sidecar containers enter the `Terminated` state and are no longer restarted if they fail.
+3. Once the last primary container terminates, the last started sidecar container is notified by
+   sending a `SIGTERM` signal, even if `PreStop` hook hasn't completed yet.
```

Comment:
This is a change from the [currently documented behavior](https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#container-hooks) of PreStop: "the hook must complete before the TERM signal to stop the container can be sent".

Have we considered keeping the behavior consistent across container types (init and main)? We can still terminate based on pod termination grace period.

Reply:
Good catch, that's actually how I implemented it in https://github.com/kubernetes/kubernetes/pull/120620 .

Reply:
As discussed today at Sidecar WG meeting, we need to be careful to make sure the sidecar scenarios can be implemented with the minima risk. We want to give a signal at the beginning of termination AND we want another signal to notify when all main containers are gone.

As written the thinking was that sidecars will follow the general pattern:

1. On `PreStop` start termination procedures, but must function. 
   a. Logging/metrics will flush buffers and move into mode of operation with the minimal buffering
   b. Service mesh will signal proxy to start draining (draining == encourage clients to send elsewhere with GOAWAY/connection: close, etc)
2. On `SIGTERM` make the final clean ups ASAP.
  a. Logging/metrics will flush the very small buffer they have left and exit as fast as it can
  b. Service mesh will exit immediately as it is not needed any longer

The idea of sending the `SIGTERM` during the `PreStop` was that implementation-wise, sidecars will implement the main clean up logic in OreStop hook itself and will have a `SIGTERM` signal to be used to interrupt whatever `PreStop` was doing.

The alternative implementation that satisfy both scenarios, while keeping the semantic of uninterrupteable `PreStop` would be for sidecar to use `PreStop` as a notification mechanism only. So sidecar implementation may be as following:

- If sidecar implemented PreStop's clean up synchronous, the whole PreStop will run and SIGTERM will be sent to the Container after both: PreStop has completed AND all main containers are gone. There may be a gap between the PreStop completion and  SIGTERM. And there may be a situation when the sidecar extended the pod lifetime to the duration of it's PreStop execution.
- Sidecar can choose to implement PreStop as a notification mechanism. PreStop starts the termination in separate thread (let's call it `PreStopAsync`) and exits the callback immediately. The `PreStopAsync` and `SIGTERM` will behave the same as written currently in the document - `SIGTERM` can interrupt the `PreStopAsync`.

The downside of this approach is that:
1. There is a change in behavior - now there is potential gap between the `PreStop` completion and the `SIGTERM`. Authors of sidecars will need to implement container to be fully functional after `PreStop` has finished.
2. Even if PreStop is fast and only starts the `PreStopAsync`, it is still an overhead. For extra-fast Pods that needs to terminate immediately, `SIGTERM` interrupting the `PreStop` will work faster.

WIth all this said, I feel like breaking the behavior by interrupting `PreStop` sounds like a better alternative. @jpbetz @tzneal wdyt?

Reply:
> There is a change in behavior - now there is potential gap between the PreStop completion and the SIGTERM

This is not a "change" - it's true today, just less noticeable.

I'm not seeing why we would want different semantics for this, but I am not sure I understand the proposal in full..

Reply:
> > There is a change in behavior - now there is potential gap between the PreStop completion and the SIGTERM
> 
> This is not a "change" - it's true today, just less noticeable.

From implementation perspective it is pretty much immediate today. So some implementaiton might take dependency on it. After discusssion at today meeting we decided to go with it though. It will be easier for people to implement.

> I'm not seeing why we would want different semantics for this, but I am not sure I understand the proposal in full..

The main motivation here is to make sure that scenarios of spot instances with the limited graceful termination time are supported as well as we can. So we want to start termination of sidecars as soon as we can so in case of logging - they have time to flush buffers. We are talking about 15 seconds total graceful termination period here.

Ae also want to make sure sidecars are working while main containers are working, but not much longer. So we need to move the `SIGTERM` to the place that is more appropriate for sidecars.

Reply:
> After discusssion at today meeting we decided to go with it though. It will be easier for people to implement.

(just to check if I understand correctly)
The sidecar container will get SIGTERM after it completes the preStop hook or the TGPS expires.
And users can implement their preStop hook in whatever way they want (synchronously or asynchronously), right?